From: Linus Torvalds <torvalds@osdl.org>
Date: Tue, 29 Nov 2005 10:38:30 +0000 (+0100)
Subject: [PATCH] Fix ptrace self-attach rule (2.6 backport)
X-Git-Tag: v2.4.33-pre1
X-Git-Url: http://www.kernel.org/git/?p=linux/kernel/git/marcelo/linux-2.4.git;a=commitdiff;h=6eaf7c8bfa4df71040936323f04ec145f095b1ea

[PATCH] Fix ptrace self-attach rule (2.6 backport)

Patch-mainline: v2.6.14.2
Acked-by: Karsten Keil <kkeil@suse.de>
X-Git-Url: http://www.kernel.org/git/?p=linux/kernel/git/gregkh/linux-2.6.14.y.git;a=commitdiff;h=082d52c56f642d21b771a13221068d40915a1409

  [PATCH] Fix ptrace self-attach rule

  Before we did CLONE_THREAD, the way to check whether we were attaching
  to ourselves was to just check "current == task", but with CLONE_THREAD
  we should check that the thread group ID matches instead.

  Signed-off-by: Linus Torvalds <torvalds@osdl.org>
---

--- a/kernel/ptrace.c
+++ b/kernel/ptrace.c
@@ -58,7 +58,7 @@ int ptrace_attach(struct task_struct *ta
 	task_lock(task);
 	if (task->pid <= 1)
 		goto bad;
-	if (task == current)
+	if (task->tgid == current->tgid)
 		goto bad;
 	if (!task->mm)
 		goto bad;
